proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,332 @@
+  for each NLP task and register them with MultiTaskComponent. The burden of defining templates for each task
+  would be on the prompt engineers and not on the users.
+
+
+- The use of templates to define NLP tasks can make it easier for users to use MultiTaskComponent, as
+  they do not need to know the details of how the model works or how to define tasks for it. This can
+  reduce the learning curve and make it easier for users to get started with MultiTaskComponent and
+  to leverage the power of LLMs in Haystack.
+
+
+- The extensible and modular approach of MultiTaskComponent allows users to easily add support for
+  additional tasks, even on-the-fly, which can enable them to extend the capabilities of the model and to use it for
+  a wider range of NLP tasks. This can provide users with more flexibility and control over the model,
+  and can enable them to tailor it to their specific needs and applications.
+
+  # Detailed design
+
+- The MultiTaskComponent class is the most important abstraction in Haystack's large language model (LLM) support.
+  In addition to MultiTaskComponent class, we'll also define a set of templates for the most popular NLP tasks.
+
+
+- NLP task templates will be represented by `TaskTemplate` dataclass class.
+
+  ``` python
+      @dataclass(frozen=True)
+      class TaskTemplate:
+
+          name: str
+          prompt_text: str
+          prompt_params: List[str]
+
+  ````
+
+  MultiTaskComponent defines 10-20 default NLP tasks with TaskTemplate instance list and allow users to register
+  additional tasks with MultiTaskComponent.
+
+
+- The default tasks (question-answering,question-generation, summarization etc.) could be examined by the user
+  using `supported_tasks_names` method. For example:
+
+  ``` python
+	  from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      mtc.supported_task_names()
+
+      ----------------------------------------------------------------------------
+      ['question-answering',
+       'question-generation',
+       'conditioned-question-generation',
+       'summarization',
+       'question-answering-check']
+  ```
+
+
+- MultiTaskComponent supports natural language prompting (using `prompt_for_task`) by specifying task in addition to
+  prompt. For example:
+
+  ``` python
+      from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+	  mtc.prompt_for_task("question-generation", documents=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ['What is the capital of Germany?']
+  ```
+
+- MultiTaskComponent supports selecting a particular current task (e.g. question-generation) and then subsequently
+  using the selected task until user changes the current task. For example:
```

Comment:
I'm not a huge fan of having the `MultiTaskComponent` having state (or encouraging the usage of state). This could lead to hard to find bugs in code when using `MultiTaskComponent` in multiple places. I think I would prefer if we only use invocations like `mtc.prompt_for_task("question-generation", documents=["Berlin is the capital of Germany."])` where the template in question is explicitly asked for.

Reply:
I see rather advantages of having an explicit state here:
- it would simplify the config of a pipeline in a yaml and make the config more transparent. Think of running experiments with different prompts. It will be easier to compare in hindsight if the prompt is treated as a param. Similar to a model name in a reader or the temperature in a generator node. 
- it's in line with haystack's design of setting most params at init time and only optionally overriding them at runtime if needed (e.g. top_k)

Reply:
I would be fine with passing it as an __init__ parameter as we do with the model parameter. But I would then not encourage users to change the parameter after initialization. In the same way we don't change the model parameter after initializing a reader node. What do you think?

Reply:
It doesn't you are right but we never had a component that can do 100+ tasks and locking it up for one task is also not ideal